### PR TITLE
[rdf-validate-shacl] Add generics

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Adds generics, so that the dataset/types in the `ValidationReport` match the given factory.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

### CODE REVIEW

File: types/rdf-validate-shacl/rdf-validate-shacl-tests.ts [hunk 1]
```diff
@@ -1,23 +1,55 @@
...
+// $ExpectType boolean
+report1.conforms;
+
+for (const result of report1.results) {
+    result.message; // $ExpectType Quad_Object[]
+    result.path; // $ExpectType Quad_Object | null
+    result.focusNode; // $ExpectType Quad_Object | null
+    result.severity; // $ExpectType Quad_Object | null
+    result.sourceConstraintComponent; // $ExpectType Quad_Object | null
+    result.sourceShape; // $ExpectType Quad_Object | null
```
Comment (thewilkybarkid) [inline types/rdf-validate-shacl/rdf-validate-shacl-tests.ts:30]: TS 4.1 complains with:

```
ERROR: 26:5  expect  TypeScript@4.1 expected type to be:
  Quad_Object | null
got:
  Quad | BlankNode | Literal | Variable | NamedNode<string> | null
```

(`Quad_Object` is an alias for `Quad | BlankNode | Literal | Variable | NamedNode<string>`, so it's compatible.)

I can add `// Minimum TypeScript Version: 4.4`, but seems a bit heavy-handed given it does work with 4.1?

File: types/rdf-validate-shacl/src/validation-report.d.ts [hunk 2]
```diff
@@ -1,30 +1,36 @@
...
+    interface ValidationResult<D extends DatasetCore> {
+        term: SubjectOf<D>;
+        dataset: D;
+        cf: GraphPointer<SubjectOf<D>, D>;
+        readonly message: Array<ObjectOf<D>>;
+        readonly path: ObjectOf<D> | null;
+        readonly focusNode: ObjectOf<D> | null;
+        readonly severity: ObjectOf<D> | null;
+        readonly sourceConstraintComponent: ObjectOf<D> | null;
+        readonly sourceShape: ObjectOf<D> | null;
```
Comment (thewilkybarkid) [inline types/rdf-validate-shacl/src/validation-report.d.ts:24]: These types are expanded but do now match `resultQuads`.

File: types/rdf-validate-shacl/src/validation-report.d.ts [hunk 3]
```diff
@@ -1,30 +1,36 @@
...
     }
 }
 
-declare class ValidationReport {
-    constructor(resultQuads: Quad[], options: ValidationReport.Options);
-    term: BlankNode | NamedNode;
-    dataset: DatasetCore;
+declare class ValidationReport<D extends DatasetCore> {
```
Comment (tpluscode) [inline types/rdf-validate-shacl/src/validation-report.d.ts:28]: Please add default for compatibility

```suggestion
declare class ValidationReport<D extends DatasetCore = DatasetCore> {
```

File: types/rdf-validate-shacl/src/validation-report.d.ts [hunk 4]
```diff
@@ -1,30 +1,36 @@
-import { BlankNode, DataFactory, DatasetCore, DatasetCoreFactory, NamedNode, Quad, Term } from 'rdf-js';
+import { DataFactory, DatasetCore, DatasetCoreFactory } from 'rdf-js';
 import { GraphPointer } from 'clownface';
 
+type FactoryFor<D> = D extends DatasetCore<infer OutQuad, infer InQuad> ? DataFactory<OutQuad, InQuad> & DatasetCoreFactory<OutQuad, InQuad, D> : never;
+type OutQuadOf<D> = D extends DatasetCore<infer Q, any> ? Q : never;
+type InQuadOf<D> = D extends DatasetCore<any, infer Q> ? Q : never;
+type SubjectOf<D> = OutQuadOf<D>['subject'];
+type ObjectOf<D> = OutQuadOf<D>['object'];
```
Comment (tpluscode) [inline types/rdf-validate-shacl/src/validation-report.d.ts:8]: Are those really necessary? Typically types from `@rdf-js/types` should suffice

File: types/rdf-validate-shacl/src/validation-report.d.ts [hunk 5]
```diff
@@ -1,30 +1,36 @@
...
-        dataset: DatasetCore;
-        cf: GraphPointer;
-        readonly message: Term[];
-        readonly path: BlankNode | NamedNode | null;
-        readonly focusNode: BlankNode | NamedNode | null;
-        readonly severity: NamedNode | null;
-        readonly sourceConstraintComponent: BlankNode | NamedNode | null;
-        readonly sourceShape: BlankNode | NamedNode | null;
+    interface ValidationResult<D extends DatasetCore> {
+        term: SubjectOf<D>;
```
Comment (tpluscode) [inline types/rdf-validate-shacl/src/validation-report.d.ts:16]: Also note that `SubjectOf<D>` is not equivalent to `BlankNode | NamedNode`...

Same goes for `ObjectOf<>`. They both replace different kinds of types and unions

File: types/rdf-validate-shacl/src/data-factory.d.ts [hunk 6]
```diff
@@ -1,12 +0,0 @@
...
-// tslint:disable-next-line no-empty-interface
-interface DataFactory extends RDF.DataFactory, RDF.DatasetCoreFactory {
-}
-
-declare class DataFactory {
-    constructor(impl?: RDF.DataFactory);
-    readonly true: RDF.Literal;
-}
-
-export = DataFactory;
```
Comment (tpluscode) [inline types/rdf-validate-shacl/src/data-factory.d.ts:12]: Doubt that this one is at all important but technically you changed the interface of `SHACLValidator#factory`, effectively removing the `true` property :)

File: types/rdf-validate-shacl/index.d.ts [hunk 7]
```diff
@@ -3,23 +3,25 @@
...
+    interface Options<F extends DataFactory & DatasetCoreFactory> {
+        factory?: F | undefined;
         maxErrors?: number | undefined;
     }
 }
 
-declare class SHACLValidator {
-    constructor(shapes: RDF.DatasetCore, options?: SHACLValidator.Options);
-    factory: DataFactory;
+declare class SHACLValidator<D extends DatasetCore, F extends FactoryFor<D>> {
```
Comment (tpluscode) [inline types/rdf-validate-shacl/index.d.ts:19]: Here too please add default types so that the interface can be used without type arguments

Also, it appears to be irrelevant whether the factory and dataset are related. I think the `FactoryFor<>` redundant

```suggestion
declare class SHACLValidator<D extends DatasetCore = DatasetCore, F extends DataFactory & DatasetCoreFactory = DataFactory & DatasetCoreFactory> {
```

File: types/rdf-validate-shacl/index.d.ts [hunk 8]
```diff
@@ -3,23 +3,25 @@
...
+    constructor(shapes: D, options?: SHACLValidator.Options<F>);
```
Comment (tpluscode) [inline types/rdf-validate-shacl/index.d.ts:20]: I propose to keep as it was

```suggestion
    constructor(shapes: DatasetCore, options?: SHACLValidator.Options<F>);
```

File: types/rdf-validate-shacl/index.d.ts [hunk 9]
```diff
@@ -3,23 +3,25 @@
...
+    factory: F;
     depth: number;
-    validate(data: RDF.DatasetCore): ValidationReport;
-    nodeConformsToShape(focusNode: RDF.Term, shapeNode: RDF.Term): boolean;
+    validate(data: D): ValidationReport<D>;
```
Comment (tpluscode) [inline types/rdf-validate-shacl/index.d.ts:23]: Ditto, because `input dataset !== output dataset`

```suggestion
    validate(data: DatasetCore): ValidationReport<D>;
```

**Review by tpluscode — CHANGES_REQUESTED** [review]
OK, so first I suggest that you remove the `SubjectOf` and `ObjectOf`. I think the node interface is not super important to match the dataset. Just adds complexity and the build problem should disappear too. In fact, much of the generic `infer` magic can be removed to simplify matters 

Second, I notice that that you made the constructor parameter the same D as the one in `validate` method. This is not correct. I think it should be allowed to provide a different type of dataset.

Correct me if I'm wrong, I assume the you need the factory type to get the correct type of `ValidationResult#dataset`.

Bottom line, most of the constructor parameters are also unnecessary and all you really need is the ability to type `SHACLValidator<DatasetExt>`. Even the may generic for factory may not necessary

File: types/rdf-validate-shacl/src/data-factory.d.ts (hunk 6 above)
Comment (thewilkybarkid) [reply types/rdf-validate-shacl/src/data-factory.d.ts:12]: Was confused by it: looks to have been removed https://github.com/zazuko/rdf-validate-shacl/pull/62.

Comment (tpluscode) [reply types/rdf-validate-shacl/src/data-factory.d.ts:12]: Sorry, I have not checked the latest 😬

File: types/rdf-validate-shacl/rdf-validate-shacl-tests.ts (hunk 1 above)
Comment (thewilkybarkid) [reply types/rdf-validate-shacl/rdf-validate-shacl-tests.ts]: Learnt from #54195 how to get round this.

**Review by tpluscode — COMMENTED** [review]
Looks good now and does not break my projects 👌 

But please explain, is it necessary to also make the term types generic? I hardly ever need the `(Term)Ext` types. The core types are just enough. Only the `DatasetExt` has real day to day value

**Comment by thewilkybarkid** [comment]
> But please explain, is it necessary to also make the term types generic? I hardly ever need the `(Term)Ext` types. The core types are just enough. Only the `DatasetExt` has real day to day value

I don't disagree that the dataset type is the most useful, and I do prefer functions instead of extended types, but I included them for completeness. (Not sure I understand the need for a `DataFactory` type otherwise.)

**Comment by tpluscode** [comment]
Ok, let's keep it. It's not gonna hurt us I suppose. I'm just not super fond of the additional complexity

> (Not sure I understand the need for a `DataFactory` type otherwise.)

Hehe, the RDF/JS spec does not case about implementation much. It is our convenience that these type are getting more and more generic arguments. Not all equally useful :)

**Review by tpluscode — APPROVED** [review]

**Comment by thewilkybarkid** [comment]
Ready to merge
